### DISCUSSION TITLE
Allow params keyword in some cases even if it is not the last parameter

### DISCUSSION THREAD

**Discussion by gurkantuna** [dotnet/csharplang#4658] [Language Ideas]
I am aware that in the following usage, the compiler is waiting for the `params` keyword as the last parameter to find out which parameter is the beginning of the array.

``` csharp
void Foo(params string[] args, params string[] args2) {
   
}
```
Also, like the compiler, I cannot predict how to find start/end of array with the usage of as follows and I agree with it :)

``` csharp
Foo("ab","bc","cd","de");//it is not clear which parameter is passed to which array! 
```

But what if we change the type of the parameter?

``` csharp
void Foo(params string[] args, params int[] args2) {
   
}

Foo("ab", "bc", 1, 2); 
```
If you show me this, I would think the `int` array starts from the third parameter. But the truth is, although I can tell the end of the arrays, I don't expect the compiler to do that.

Above methods cannot be compiled caused by my wrong usage :
> CS0231 Error-A params parameter must be the last parameter in a formal
> parameter list

Can be compiled in the following case, but I think it's pointless to use an optional parameter since the parameter has to be passed.

``` csharp
void Bar(bool arg = false, params string[] args) {

}

Bar(false, "ab", "bc");
```
So I think the following usage is deterministic as the last parameter is already of boolean type. I mean, I just want to use the ``Bar` method like this :

``` csharp
void Bar(params string[] args, bool arg = false) {

}

Bar( "ab", "bc");
```

Maybe I'm missing something, but I think this usage should be allowed in the language.

<a name="ps">P.S.<a/> My suggestion is, of course, eligible for only the non-object types and different types of non-consecutive parameters. 

<pre><code><del>void Bar(params object[] args, bool arg = false){</del>
   
<del>}</del>

<del>void Bar(params string[] args, params string[] args2){</del>
   
<del>}</del>
</pre></code>

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/1458

**Reply by gurkantuna** [reply]
I've already mentioned it, thanks anyway.
See: [#ps](#ps)
